File: code/packages/typescript/wasm-runtime/tests/square.test.ts
/**
 * End-to-End Test: square(n) = n * n
 *
 * ==========================================================================
 * The Ultimate Integration Test
 * ==========================================================================
 *
 * This test proves the entire WASM stack works end-to-end:
 *
 *   Raw .wasm bytes → Parse → Validate → Instantiate → Execute → Result
 *
 * The test hand-assembles a minimal WASM module that exports a ``square``
 * function: ``(i32) -> (i32)`` which computes ``n * n``.
 *
 * In WAT (WebAssembly Text Format), this module is:
 *
 * ```wat
 * (module
 *   (type (func (param i32) (result i32)))
 *   (func (export "square") (type 0)
 *     local.get 0
 *     local.get 0
 *     i32.mul)
 * )
 * ```
 *
 * The binary encoding is hand-assembled byte by byte below, following the
 * WASM 1.0 binary format specification.
 *
 * @module
 */

import { describe, it, expect } from "vitest";
import { WasmRuntime } from "../src/wasm_runtime.js";
import { encodeSigned, encodeUnsigned } from "@coding-adventures/wasm-leb128";

// =========================================================================
// Hand-Assembled WASM Module: square(n) = n * n
// =========================================================================

/**
 * Build the square.wasm binary by hand.
 *
 * This is a minimal valid WASM module with:
 * - 1 type:     (i32) -> (i32)
 * - 1 function: uses type 0
 * - 1 export:   "square" -> function 0
 * - 1 code body: local.get 0, local.get 0, i32.mul, end
 */
function buildSquareWasm(): Uint8Array {
  const parts: number[] = [];

  // ── Header ──────────────────────────────────────────────────────────
  // Magic: "\0asm"
  parts.push(0x00, 0x61, 0x73, 0x6D);
  // Version: 1
  parts.push(0x01, 0x00, 0x00, 0x00);

  // ── Type Section (ID 1) ─────────────────────────────────────────────
  // Contains 1 function type: (i32) -> (i32)
  const typePayload = [
    0x01,       // 1 type entry
    0x60,       // function type marker
    0x01, 0x7F, // 1 param: i32
    0x01, 0x7F, // 1 result: i32
  ];
  parts.push(0x01);                           // section ID = 1 (Type)
  parts.push(...encodeUnsigned(typePayload.length)); // section size
  parts.push(...typePayload);

  // ── Function Section (ID 3) ─────────────────────────────────────────
  // Contains 1 function pointing to type index 0
  const funcPayload = [
    0x01,       // 1 function
    0x00,       // type index 0
  ];
  parts.push(0x03);                           // section ID = 3 (Function)
  parts.push(...encodeUnsigned(funcPayload.length));
  parts.push(...funcPayload);

  // ── Export Section (ID 7) ───────────────────────────────────────────
  // Exports "square" as function 0
  const nameBytes = new TextEncoder().encode("square");
  const exportPayload = [
    0x01,                                     // 1 export
    ...encodeUnsigned(nameBytes.length),       // name length
    ...nameBytes,                              // name bytes
    0x00,                                     // export kind: function
    0x00,                                     // function index 0
  ];
  parts.push(0x07);                           // section ID = 7 (Export)
  parts.push(...encodeUnsigned(exportPayload.length));
  parts.push(...exportPayload);

  // ── Code Section (ID 10) ────────────────────────────────────────────
  // Contains 1 function body:
  //   local.get 0    (0x20 0x00)
  //   local.get 0    (0x20 0x00)
  //   i32.mul        (0x6C)
  //   end            (0x0B)
  const bodyCode = [
    0x20, 0x00,   // local.get 0
    0x20, 0x00,   // local.get 0
    0x6C,         // i32.mul
    0x0B,         // end
  ];
  const bodyPayload = [
    0x00,         // 0 local declarations
    ...bodyCode,
  ];
  const funcBody = [
    ...encodeUnsigned(bodyPayload.length),    // body size
    ...bodyPayload,
  ];
  const codePayload = [
    0x01,         // 1 function body
    ...funcBody,
  ];
  parts.push(0x0A);                           // section ID = 10 (Code)
  parts.push(...encodeUnsigned(codePayload.length));
  parts.push(...codePayload);

  return new Uint8Array(parts);
}

// =========================================================================
// Tests
// =========================================================================

describe("square.wasm end-to-end", () => {
  it("should compute square(5) = 25", () => {
    const runtime = new WasmRuntime();
    const wasmBytes = buildSquareWasm();
    const result = runtime.loadAndRun(wasmBytes, "square", [5]);
    expect(result).toEqual([25]);
  });

  it("should compute square(0) = 0", () => {
    const runtime = new WasmRuntime();
    const wasmBytes = buildSquareWasm();
    const result = runtime.loadAndRun(wasmBytes, "square", [0]);
    expect(result).toEqual([0]);
  });

  it("should compute square(-3) = 9", () => {
    const runtime = new WasmRuntime();
    const wasmBytes = buildSquareWasm();
    const result = runtime.loadAndRun(wasmBytes, "square", [-3]);
    expect(result).toEqual([9]);
  });

  it("should handle i32 overflow: square(2147483647) wraps", () => {
    const runtime = new WasmRuntime();
    const wasmBytes = buildSquareWasm();
    const result = runtime.loadAndRun(wasmBytes, "square", [2147483647]);
    // 2147483647 * 2147483647 = 4611686014132420609, which wraps to 1 in i32
    // Math.imul(2147483647, 2147483647) = 1
    expect(result).toEqual([1]);
  });

  it("should support step-by-step: load, validate, instantiate, call", () => {
    const runtime = new WasmRuntime();
    const wasmBytes = buildSquareWasm();

    // Step 1: Parse
    const module = runtime.load(wasmBytes);
    expect(module.types.length).toBe(1);
    expect(module.functions.length).toBe(1);
    expect(module.exports.length).toBe(1);

    // Step 2: Validate
    const validated = runtime.validate(module);
    expect(validated.module).toBe(module);

    // Step 3: Instantiate
    const instance = runtime.instantiate(module);
    expect(instance.exports.has("square")).toBe(true);

    // Step 4: Call
    const result = runtime.call(instance, "square", [7]);
    expect(result).toEqual([49]);
  });
});

File: code/packages/elixir/wasm_runtime/test/square_test.exs
defmodule CodingAdventures.WasmRuntime.SquareTest do
  use ExUnit.Case, async: true

  @moduledoc """
  End-to-end test: compile a "square" function as raw WASM bytecodes,
  parse + validate + instantiate + call it, and verify square(5) == 25.

  This test constructs a minimal valid WASM module by hand (no compiler
  needed). The module exports a single function `square(x: i32) -> i32`
  that computes `x * x` using `local.get 0` twice and `i32.mul`.

  ## WASM Module Structure

      (module
        (type (func (param i32) (result i32)))    ;; type 0
        (func (export "square") (type 0)           ;; func 0
          local.get 0
          local.get 0
          i32.mul
        )
      )

  ## Binary Layout

      Header:  \\x00asm\\x01\\x00\\x00\\x00
      Type section (1):   1 type -> (i32) -> (i32)
      Function section (3): 1 function -> type index 0
      Export section (7):   "square" -> function 0
      Code section (10):    1 body -> [local.get 0, local.get 0, i32.mul, end]
  """

  alias CodingAdventures.WasmExecution.Values

  # Build a minimal WASM module binary by hand
  defp build_square_wasm do
    # -- WASM Header --
    header = <<0x00, 0x61, 0x73, 0x6D, 0x01, 0x00, 0x00, 0x00>>

    # -- Type Section (ID=1) --
    # One function type: (i32) -> (i32)
    #   0x60 = func type tag
    #   0x01 0x7F = 1 param, i32
    #   0x01 0x7F = 1 result, i32
    type_entry = <<0x60, 0x01, 0x7F, 0x01, 0x7F>>
    # count = 1
    type_section_body = <<0x01>> <> type_entry
    type_section = <<0x01>> <> leb128_size(type_section_body) <> type_section_body

    # -- Function Section (ID=3) --
    # One function referencing type index 0
    # count = 1, type_idx = 0
    func_section_body = <<0x01, 0x00>>
    func_section = <<0x03>> <> leb128_size(func_section_body) <> func_section_body

    # -- Export Section (ID=7) --
    # Export "square" as function index 0
    export_name = "square"
    export_entry = leb128_size(export_name) <> export_name <> <<0x00, 0x00>>
    # count = 1
    export_section_body = <<0x01>> <> export_entry
    export_section = <<0x07>> <> leb128_size(export_section_body) <> export_section_body

    # -- Code Section (ID=10) --
    # One function body:
    #   locals: 0 local declarations
    #   code: local.get 0, local.get 0, i32.mul, end
    func_code = <<0x20, 0x00, 0x20, 0x00, 0x6C, 0x0B>>
    # 0 local decl groups + code
    func_body = <<0x00>> <> func_code
    func_body_with_size = leb128_size(func_body) <> func_body
    # count = 1
    code_section_body = <<0x01>> <> func_body_with_size
    code_section = <<0x0A>> <> leb128_size(code_section_body) <> code_section_body

    header <> type_section <> func_section <> export_section <> code_section
  end

  # Encode a byte count as a single-byte LEB128 (sufficient for small values)
  defp leb128_size(data) when is_binary(data), do: <<byte_size(data)>>

  test "square(5) returns 25" do
    wasm_bytes = build_square_wasm()

    # Parse
    {:ok, wasm_module} = CodingAdventures.WasmModuleParser.parse(wasm_bytes)
    assert length(wasm_module.types) == 1
    assert length(wasm_module.functions) == 1
    assert length(wasm_module.code) == 1

    # Validate
    {:ok, validated} = CodingAdventures.WasmValidator.validate(wasm_module)
    assert length(validated.func_types) == 1

    # Instantiate
    {:ok, instance} = CodingAdventures.WasmRuntime.Instance.from_validated(validated)

    # Call square(5)
    results = CodingAdventures.WasmRuntime.Instance.call(instance, "square", [Values.i32(5)])

    assert length(results) == 1
    [result] = results
    # i32
    assert result.type == 0x7F
    assert result.value == 25
  end

  test "square(0) returns 0" do
    wasm_bytes = build_square_wasm()
    {:ok, wasm_module} = CodingAdventures.WasmModuleParser.parse(wasm_bytes)
    {:ok, validated} = CodingAdventures.WasmValidator.validate(wasm_module)
    {:ok, instance} = CodingAdventures.WasmRuntime.Instance.from_validated(validated)

    [result] = CodingAdventures.WasmRuntime.Instance.call(instance, "square", [Values.i32(0)])
    assert result.value == 0
  end

  test "square(-3) returns 9" do
    wasm_bytes = build_square_wasm()
    {:ok, wasm_module} = CodingAdventures.WasmModuleParser.parse(wasm_bytes)
    {:ok, validated} = CodingAdventures.WasmValidator.validate(wasm_module)
    {:ok, instance} = CodingAdventures.WasmRuntime.Instance.from_validated(validated)

    [result] = CodingAdventures.WasmRuntime.Instance.call(instance, "square", [Values.i32(-3)])
    assert result.value == 9
  end

  test "Runtime.instantiate_bytes convenience API" do
    wasm_bytes = build_square_wasm()
    {:ok, instance} = CodingAdventures.WasmRuntime.Runtime.instantiate_bytes(wasm_bytes)

    results = CodingAdventures.WasmRuntime.Runtime.call(instance, "square", [Values.i32(7)])
    assert [%{value: 49}] = results
  end
end
